server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/TagAggregationOperator.java

```diff
@@ -0,0 +1,237 @@
+  private final int[] consumedIndices;
+  private final BitMap dataReady;
+  private final BitMap childFinished;
+  private final TsBlockBuilder tsBlockBuilder;
+  private final long maxRetainedSize;
+  private final long childrenRetainedSize;
+  private final long maxReturnSize;
+
+  public TagAggregationOperator(
+      OperatorContext operatorContext,
+      List<List<String>> groups,
+      List<List<Aggregator>> groupedAggregators,
+      List<Operator> children,
+      long maxReturnSize) {
+    this.operatorContext = Validate.notNull(operatorContext);
+    this.groups = Validate.notNull(groups);
+    this.groupedAggregators = Validate.notNull(groupedAggregators);
+    this.children = Validate.notNull(children);
+    List<TSDataType> actualOutputColumnTypes = new ArrayList<>();
+    for (String ignored : groups.get(0)) {
+      actualOutputColumnTypes.add(TSDataType.TEXT);
+    }
+    for (int outputColumnIdx = 0;
+        outputColumnIdx < groupedAggregators.get(0).size();
+        outputColumnIdx++) {
+      for (List<Aggregator> aggregators : groupedAggregators) {
+        Aggregator aggregator = aggregators.get(outputColumnIdx);
+        if (aggregator != null) {
+          actualOutputColumnTypes.addAll(Arrays.asList(aggregator.getOutputType()));
+          break;
+        }
+      }
+    }
+    this.tsBlockBuilder = new TsBlockBuilder(actualOutputColumnTypes);
+    // Initialize input tsblocks for each aggregator group.
+    this.inputTsBlocks = new TsBlock[children.size()];
+    this.consumedIndices = new int[children.size()];
+    this.dataReady = new BitMap(children.size());
+    this.childFinished = new BitMap(children.size());
+    this.maxRetainedSize = children.stream().mapToLong(Operator::calculateMaxReturnSize).sum();
+    this.childrenRetainedSize =
+        children.stream().mapToLong(Operator::calculateRetainedSizeAfterCallingNext).sum();
+    this.maxReturnSize = maxReturnSize;
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public TsBlock next() {
+    long maxRuntime = operatorContext.getMaxRunTime().roundTo(TimeUnit.NANOSECONDS);
+    long start = System.nanoTime();
+    boolean successful = true;
+    while (System.nanoTime() - start < maxRuntime && !tsBlockBuilder.isFull() && successful) {
+      successful = processOneRow();
+    }
+    TsBlock tsBlock = null;
+    if (tsBlockBuilder.getPositionCount() > 0) {
+      tsBlock = tsBlockBuilder.build();
+    }
+    tsBlockBuilder.reset();
+    return tsBlock;
+  }
+
+  private boolean processOneRow() {
+    for (int i = 0; i < children.size(); i++) {
+      if (dataReady.isMarked(i)) {
+        continue;
+      }
+      if (dataUnavailable(i)) {
+        // Find next non-empty tsblock of the child
+        do {
+          inputTsBlocks[i] = children.get(i).next();
+        } while (inputTsBlocks[i] != null && inputTsBlocks[i].isEmpty());
```

Comment:
we should only call child.next() only once in each operator.next(), the reason can be found in `Operator implementation` chapter of this [feishu doc](https://apache-iotdb.feishu.cn/docs/doccnWnRJQmCWqscG0sO3V4BMnd#WzO22U)

Reply:
Fixed.